### PR TITLE
558 fix resubmission error two electric boogaloo

### LINKED ISSUE #558: Properly Serialize NULL Instant's
## Overview

`NULL` has to be carefully treated as the absence of values within the `CommitVerification` system is intentionally used with meaning.

## Root Cause

The error output from the issues.

### Original Issue

A student posted this image which was likely caused by this issue:

image

Backend stack traces:

```
edu.byu.cs.dataAccess.DataAccessException: Error inserting item into table submission
        at edu.byu.cs.dataAccess.sql.helpers.SqlReader.insertItem(SqlReader.java:128)
        at edu.byu.cs.dataAccess.sql.SubmissionSqlDao.insertSubmission(SubmissionSqlDao.java:89)
        at edu.byu.cs.autograder.Grader.run(Grader.java:106)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: edu.byu.cs.util.Serializer$SerializationException: java.lang.NullPointerException: Cannot invoke "java.time.Instant.toString()" because "instant" is null
        at edu.byu.cs.util.Serializer.serialize(Serializer.java:30)
        at edu.byu.cs.model.Submission.serializeObject(Submission.java:183)
        at edu.byu.cs.model.Submission.serializeCommitResult(Submission.java:179)
        at edu.byu.cs.dataAccess.sql.helpers.SqlReader.setValue(SqlReader.java:133)
        at edu.byu.cs.dataAccess.sql.helpers.SqlReader.insertItem(SqlReader.java:123)
        ... 7 common frames omitted
Caused by: java.lang.NullPointerException: Cannot invoke "java.time.Instant.toString()" because "instant" is null
        at edu.byu.cs.util.Serializer$InstantAdapter.write(Serializer.java:124)
        at edu.byu.cs.util.Serializer$InstantAdapter.write(Serializer.java:120)
        at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.write(TypeAdapterRuntimeTypeWrapper.java:70)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.write(ReflectiveTypeAdapterFactory.java:196)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.write(ReflectiveTypeAdapterFactory.java:368)
        at com.google.gson.Gson.toJson(Gson.java:842)
        at com.google.gson.Gson.toJson(Gson.java:812)
        at com.google.gson.Gson.toJson(Gson.java:759)
        at com.google.gson.Gson.toJson(Gson.java:736)
        at edu.byu.cs.util.Serializer.serialize(Serializer.java:28)
        ... 11 common frames omitted
```

_Originally posted by @19mdavenport in https://github.com/softwareconstruction240/autograder/pull/556#issuecomment-2691831715_

### Cascading Issue

Additionally, this code caused another error by assuming that `commitResult()` would never be NULL.

image

In this case, the error is especially painful because it also causes an error when attempting to manually approve the submission:

image

_Originally posted by @frozenfrank in https://github.com/softwareconstruction240/autograder/issues/556#issuecomment-2692363139_

### PR DESCRIPTION
**Pull request by 19mdavenport**
## Overview
A better fix than #556 with fewer errors

## Details
Reverts the previous fix and replaces with a better, more permanent solution

## Testing

- [ ] Added/updated unit tests
- [ ] Tested edge cases
- [x] Manual testing (if needed)

## Additional Notes

Resolves #558

### CODE REVIEW

File: src/main/java/edu/byu/cs/util/Serializer.java [hunk 1]
```diff
@@ -117,16 +102,41 @@ public CanvasRubricAssessment deserialize(JsonElement jsonElement, Type type,
...
     }
 
-    private static class InstantAdapter extends TypeAdapter<Instant> {
+    private abstract static class NullSafeTypeAdapter<T> extends TypeAdapter<T> {
+        @Override
+        public void write(JsonWriter jsonWriter, T t) throws IOException {
+            jsonWriter.value(t == null ? null : writeNotNull(t));
+        }
+
+        private String writeNotNull(@NotNull T t) {
```
Comment (frozenfrank) [inline src/main/java/edu/byu/cs/util/Serializer.java:111]: Shouldn't this be `protected`?

File: src/main/java/edu/byu/cs/util/Serializer.java [hunk 2]
```diff
@@ -117,16 +102,41 @@ public CanvasRubricAssessment deserialize(JsonElement jsonElement, Type type,
         }
     }
 
-    private static class InstantAdapter extends TypeAdapter<Instant> {
+    private abstract static class NullSafeTypeAdapter<T> extends TypeAdapter<T> {
```
Comment (frozenfrank) [inline src/main/java/edu/byu/cs/util/Serializer.java:105]: Thank you for extracting this functionality into a class to reuse the functionality

**Review by frozenfrank — COMMENTED** [review]
Thanks for doing this, @19mdavenport! You're amazing. Thank you for you dedicated efforts to build and maintain the best autograder on campus.

I'll adjust the privacy of the method when I add the tests
